identity/src/main/java/com/android/identity/L2CAPClient.java

```diff
@@ -0,0 +1,234 @@
+            reportError(new Error("Error writing message on L2CAP socket", e));
+            return;
+        }
+
+        if (mLog.isTransportEnabled()) {
+            mLog.transport("Writing L2CAP socket");
+        }
+        while (isConnected()) {
+            byte[] messageToSend;
+            try {
+                messageToSend = mWriterQueue.poll(1000, TimeUnit.MILLISECONDS);
+                if (messageToSend == null) {
+                    continue;
+                }
+            } catch (InterruptedException e) {
+                continue;
+            }
+
+            Log.d(TAG, "Writing (" + messageToSend.length + " bytes) on L2CAP socket");
+
+            try {
+                os.write(messageToSend);
+                os.flush();
+
+            } catch (IOException e) {
+                Log.e(TAG, "Error writing message on L2CAP socket", e);
+                reportError(new Error("Error writing message on L2CAP socket", e));
+                return;
+            }
+        }
+    }
+
+    private void readFromSocket() {
+        if (mLog.isTransportEnabled()) {
+            mLog.transport("Start reading socket input");
+        }
+        // Use the max size as buffer for L2CAP socket
+        byte[] mmBuffer = new byte[mSocket.getMaxReceivePacketSize()];
+        int numBytes; // bytes returned from read()
+        // Keep listening to the InputStream until an exception occurs.
+        while (isConnected()) {
+            try {
+                // Read from the InputStream.
+                numBytes = mSocket.getInputStream().read(mmBuffer);
+                // Returns -1 if there is no more data because the end of the stream has been reached.
+                if (numBytes == -1) {
+                    if (mLog.isTransportEnabled()) {
+                        mLog.transport("End of stream reading from socket");
+                    }
+                    reportPeerDisconnected();
+                    break;
+                }
+                if (mLog.isTransportVerboseEnabled()) {
+                    Util.dumpHex(TAG, "Chunk received by socket: (" + numBytes + ")", Arrays.copyOf(mmBuffer, numBytes));
+                }
+                // Report message received.
+                mIncomingMessage.write(mmBuffer, 0, numBytes);
+                byte[] entireMessage = mIncomingMessage.toByteArray();
+                int size = Util.cborGetLength(entireMessage);
+                // Check last chunk received
+                // - if bytes received are less than buffer
+                // - or message length is equal as expected size of the message
+                if (numBytes < mmBuffer.length || (size != -1 && entireMessage.length == size)) {
+                    if (mLog.isTransportEnabled()) {
+                        mLog.transport("Data size from message: (" + size + ") message size: (" + entireMessage.length + ")");
+                    }
+                    mIncomingMessage.reset();
```

Comment:
Consider this example: 
- mmBuffer is 64 KiB
- sender sends 100 KiB worth of CBOR then 50 KiB worth of CBOR

Instead of `ByteArrayOutputStream`, consider maybe using another datastructure for `mIncomingMessage` ... for example some kind of queue or stack.

Reply:
done